File: Solutions/1986. Minimum Number of Work Sessions to Finish the Tasks/1986. Minimum Number of Work Sessions to Finish the Tasks.go

func encode(sessions *[]int ,pos int) string{
    copys:=make([]int,len(*sessions))
    copy(copys,*sessions)
    sort.Ints(copys)
    key:=strconv.Itoa(pos)+"$"
    fmt.Printf(key)
    for i:=0;i<len(copys);i+=1{
        key+=strconv.Itoa(copys[i])+"$"
    }
    return key
}

func solve(um map[string]int,sessions,tasks []int, n,sessionTime,pos int) int {
    if pos>=n{
        return 0
    }
    key:=encode(&sessions,pos)
    if val,ok:=um[key];ok{
        return val
    }
    sessions=append(sessions,tasks[pos])
    res:=1+solve(um,sessions,tasks,n,sessionTime,pos+1)
    sessions=sessions[:len(sessions)-1]
    for i:=0;i<len(sessions);i+=1{
        if sessions[i]+tasks[pos]<=sessionTime{
            sessions[i]+=tasks[pos]
            res=min(res,solve(um,sessions,tasks,n,sessionTime,pos+1))
            sessions[i]-=tasks[pos]
        }
    }
    um[key]=res
    return res
}

func min(i,j int)int{
    if i<j{
        return i
    }
    return j
}

func minSessions(tasks []int, sessionTime int) int {
    var sessions []int
    var um=make(map[string]int)
    return solve(um,sessions,tasks,len(tasks),sessionTime,0)
}

File: Solutions/1986. Minimum Number of Work Sessions to Finish the Tasks/1986. Minimum Number of Work Sessions to Finish the Tasks.cpp
class Solution {
public:
    vector<int> sessions;
    unordered_map<string,int> um;
    string encode(vector<int>& sessions,int pos){
        vector<int> copy=sessions;
        sort(copy.begin(),copy.end());
        string key=to_string(pos)+"$";
        cout << key ;
        for(int i=0;i<copy.size();++i){
            key+=to_string(copy[i])+"$";
        }
        return key;
    }
    int solve(vector<int>& tasks,int n,int sessionTime,int pos){
        if(pos>=n){
            return 0;
        }
        string key=encode(sessions,pos);
        if(um.find(key)!=um.end()){
            return um[key];
        }
        sessions.push_back(tasks[pos]);
        int res=1+solve(tasks,n,sessionTime,pos+1);
        sessions.pop_back();
        for(int i=0;i<sessions.size();++i){
            if(sessions[i]+tasks[pos]<=sessionTime){
                sessions[i]+=tasks[pos];
                res=min(res,solve(tasks,n,sessionTime,pos+1));
                sessions[i]-=tasks[pos];
            }
        }
        return um[key]=res;
    }
    int minSessions(vector<int>& tasks, int sessionTime) {
        return solve(tasks,tasks.size(),sessionTime,0);
    }
};